src/cljs/ui/websocket_fx.cljs

```diff
@@ -0,0 +1,220 @@
+                       (fn [{:keys [db]} [_ socket-id request-id & more]]
+                         (let [path    [::sockets socket-id :requests request-id]
+                               request (get-in db path)]
+                           (cond-> {:db (dissoc-in db path)}
+                             (contains? request :on-timeout)
+                             (assoc :dispatch (concatv (:on-timeout request) more))))))
+
+;; SUBSCRIPTIONS
+
+(re-frame/reg-event-fx ::subscribe
+                       (fn [{:keys [db]} [_ socket-id topic {:keys [message] :as command}]]
+                         (let [path    [::sockets socket-id :subscriptions topic]
+                               payload {:id topic :proto :subscription :data message}]
+                           {:db          (assoc-in db path command)
+                            ::ws-message {:socket-id socket-id :message payload}})))
+
+(re-frame/reg-event-fx ::subscription-message
+                       (fn [{:keys [db]} [_ socket-id subscription-id & more]]
+                         (let [path         [::sockets socket-id :subscriptions subscription-id]
+                               subscription (get-in db path)]
+                           (cond-> {}
+                             (contains? subscription :on-message)
+                             (assoc :dispatch (concatv (:on-message subscription) more))))))
+
+(re-frame/reg-event-fx ::unsubscribe
+                       (fn [{:keys [db]} [_ socket-id subscription-id & more]]
+                         (let [path         [::sockets socket-id :subscriptions subscription-id]
+                               payload      {:id subscription-id :proto :subscription :close true}
+                               subscription (get-in db path)]
+                           (cond-> {:db (dissoc-in db path)}
+                             (some? subscription)
+                             (assoc ::ws-message {:socket-id socket-id :message payload})
+                             (contains? subscription :on-close)
+                             (assoc :dispatch (concatv (:on-close subscription) more))))))
+
+(re-frame/reg-event-fx ::subscription-closed
+                       (fn [{:keys [db]} [_ socket-id subscription-id & more]]
+                         (let [path [::sockets socket-id :subscriptions subscription-id]]
+                           (if-some [subscription (get-in db path)]
+                             (cond-> {:db (dissoc-in db path)}
+                               (contains? subscription :on-close)
+                               (assoc :dispatch (concatv (:on-close subscription) more)))))))
+
+;; PUSH
+
+(re-frame/reg-event-fx ::push
+                       (fn [_ [_ socket-id command]]
+                         (let [payload {:id (random-uuid) :proto :push :data command}]
+                           {::ws-message {:socket-id socket-id :message payload}})))
+
+;; FX HANDLERS
+
+(re-frame/reg-fx
```

Comment:
I understood the implementation here but it took me a while since it is nesting go blocks and relying on closures a lot.
I also see that this implementation makes working with timeouts easier.

I went and wrote what I find is a simpler implementation, it is untested but it just uses one go block and a loop
to serially process everything that is coming from the client or the server.

I also changed one naming, which is command-ch (for lack of a better name) instead of sink-proxy, but same thing, is just the channel where clients send their intentions. I changed it just to make a distinction from the sink, since instead of sending whatever you put there it will analyze it and work with :requests or :subscriptions.
```clojure
(def connections
  "Socket ids -> command channels"
  (atom {})) 

(re-frame/reg-fx
 ::connect
 (fn [{socket-id
       :socket-id
       {:keys [url protocols on-connect on-disconnect]
        :or   {url (websocket-url)}}
       :options}]
   (async/go
     (let [command-ch (async/chan 100)
           {:keys [socket source sink close-status]} (async/<! (ws-client/connect url {:protocols protocols}))]
       (swap! connections assoc socket-id command-ch)

       
       (loop [msg-id->command {}] ;; msg-id -> [timeout command]         
         (if-not (async/poll! close-status)                     
           ;; loop grabing a message from the command-ch or the source and process it accordignly,
           ;; every second break to check for timeouts
           (let [break (async/timeout 1000) 
                 [value port] (async/alts! [command-ch source break])]
             (cond
               
               ;; user command (wants to send a request or subscribe)
               (= port command-ch)
               (let [{:keys [id proto data close timeout] :or {timeout 10000}} value]              
                 
                 ;; write the request or subscription command data on the sink ch, so it goes to the server
                 (async/>! sink (merge data {:id id :proto proto}))
                 (recur (assoc msg-id->command id [(now) proto])))
               
               ;; message from the server
               (= port source)
               (let [msg-id (:id value)
                     [_ command] (get msg-id->command msg-id)]
                 (case command
                   :request      (re-frame/dispatch [::request-response socket-id msg-id value])
                   :subscription (re-frame/dispatch [::subscription-message socket-id msg-id value])))
               
               ;; timeout check break
               (= port break)
               (let [now (now)
                     timed-outs (filter (fn [[_ [time-out _]]] (> to now)) msg-id->command)]
                 ;; notify timeouts 
                 (doseq [[msg-id [to cmd]] timed-outs]
                   (re-frame/dispatch [::request-timeout socket-id msg-id :timeout]))
                 (recur (apply dissoc msg-id->command (map first timed-outs))))))

           ;; the ws is closed, report
           (re-frame/dispatch [::disconnected socket-id closed])
           ))))))
```

Reply:
can you PR it on top of my PR? It will be easier to test it for me. You got me curious with the word "simpler" ;)

Reply:
note: agreed to create a separate PR